### ISSUE TITLE
Question: process.hrtime()

### ISSUE THREAD

**Issue by dtex** [rwaldron/temporal#17]
Is there a reason temporal doesn't use node's ```process.hrtime()``` instead of Date.now()?

I'm working on stepper motor movements and the lack of resolution accumulates. If I'm moving a stepper at 300 steps per second then I am trying to step every 3.33 milliseconds. If I tell it to set an interval that executes after 3.33ms it's not going to trigger until at least 4ms have passed so the whole movement ends up taking perceptibly longer.

**Comment by dtex** [comment]
I'm playing around with hrtime and I've encountered something strange.

I have this test that increments a variable 5000 times in a temporal loop (once every ms). In a perfect scenario this would take exactly 5 seconds to run, but of course it takes a little longer.

````js
// Using the current temporal/master
var temporal = require("../lib/temporal.js"), 
  j = 0,
  runs = 3;

runTest();

function runTest() {
  
  var startAt = process.hrtime();
  var i = 0;
  
  var queue = temporal.loop(1, function() {
    if (++i === 5000) {
      let now = process.hrtime();
      console.log("Run: " + ++j, "=", (((now[0] * NS_PER_SEC + now[1]) - startAt[0] * NS_PER_SEC + startAt[1])/NS_PER_MS));
      this.stop();
      if (j < runs) {
        process.nextTick(function() {runTest(); });
      } else {
        process.kill(process.pid, 'SIGINT');
      }
    }
  });
}
````
The weird part is that the actual time it takes to run varies between executions. If I run the test 3 time in a single execution, the times for each run will all be within 3ms of each other, but times can vary by nearly two seconds between executions. Again, they are pretty consistent within an execution of the app but vary between executions by a lot.

````bash
Donovans-MacBook-Pro:test donovan$ node speed
Run: 1 = 6370.688001
Run: 2 = 6373.639804
Run: 3 = 6373.118214

Donovans-MacBook-Pro:test donovan$ node speed
Run: 1 = 6496.228398
Run: 2 = 6499.442392
Run: 3 = 6498.856157

Donovans-MacBook-Pro:test donovan$ node speed
Run: 1 = 6860.392003
Run: 2 = 6863.192392
Run: 3 = 6862.596886

Donovans-MacBook-Pro:test donovan$ node speed
Run: 1 = 5949.319301
Run: 2 = 5952.711603
Run: 3 = 5952.172973

Donovans-MacBook-Pro:test donovan$
````

Any idea what could possibly be causing this?

**Comment by dtex** [comment]
I've been playing around with ```process.hrtime()``` and am getting encouraging results.

I have a branch of temporal that uses ```process.hrtime()``` instead of ```Date.now()``` and adds a ```temporal.resolution()``` method that lets you choose between 1ms, 0.1ms, and 0.01ms resolution.

These tests were of a temporal loop where the task ran 5000 times, and that was looped 3 times in each execution. 

````bash
// date.now() @ 1ms resolution / Running the task every 1ms
Run: 1 = 5145.196716ms
Run: 2 = 4999.54153ms
Run: 3 = 4999.865082ms

// process.hrtime() @ 1ms resolution / Running the task every 1ms
Run: 1 = 5136.149136ms
Run: 2 = 4999.568536ms
Run: 3 = 4999.900625ms

// process.hrtime() @ 0.1ms resolution / Running the task every 0.1ms
Run: 1 = 638.785268ms
Run: 2 = 502.117499ms
Run: 3 = 499.905152ms

// process.hrtime() @ 0.01ms resolution / Running the task every 0.01ms
Run: 1 = 58.600397ms
Run: 2 = 54.709429ms
Run: 3 = 51.546093ms

// process.hrtime() @ 0.01ms resolution / Running the task every 0.33ms <-- This is the cool part
Run: 1 = 1844.794861ms
Run: 2 = 1670.053478ms
Run: 3 = 1669.048576ms
````
I ran all these test under 0x and saw that processQueue spends less time on the stack with process.hrtime() vs Date.now()

When I put a counter on processQueue I found we averaged 611,433 trips through the event loop per second with ```Date.now()```and 681,426 trips through the event loop with ```process.hrtime()``` when executing the task once every ms.

Trips through the event loop remained over 600,000 per second when running the task every 0.1ms but were cut in half once I started executing the task every 0.01ms. Obviously the complexity of the task would be a big factor here.

If I set the resolution to 0.01ms but only ran the task every 1ms then performance suffered by about 10%. I assume this is because the arrays became unnecessarily large and just means a user shouldn't use the higher resolution unless they really need it.

Does this sound like a good path? I need to test in other versions of node (only used 8.7.0 so far), update tests and settle on an API for ```temporal.resolution()```. I don't want to mess with it if this is not a good idea.

**Comment by dtex** [comment]
Standard deviation from the expected times is higher when using hrTime in temporal. The longer the delay and the higher the resolution, the less accurate the timer is. In other words, accuracy is inversely proportional to the length of the queue array that gets created. 

I'm going to experiment with some alternate array structures and see if I can make it faster.

**Comment by dtex** [comment]
This landed in 0.7.0
